vllm/model_executor/layers/sampler.py

```diff
@@ -90,19 +86,48 @@ def forward(
         return _sample(probs, logprobs, input_metadata)
 
 
+def _get_logits(hidden_states: torch.Tensor, embedding: torch.Tensor,
+                embedding_bias: Optional[torch.Tensor],
+                vocab_size: int) -> torch.Tensor:
+    # Get the logits for the next tokens.
+    logits = torch.matmul(hidden_states, embedding.t())
+    if embedding_bias is not None:
+        logits += embedding_bias
+    logits = gather_from_tensor_model_parallel_region(logits)
+    # Remove paddings in vocab (if any).
+    logits = logits[:, :vocab_size]
+    return logits
+
+
 def _prune_hidden_states(
     hidden_states: torch.Tensor,
     input_metadata: InputMetadata,
 ) -> torch.Tensor:
+    last_token_indices = {t: [] for t in SamplingType}
```

Comment:
We can just have a list of lists here

Reply:
enum by default starts from 1 (if use `enum.auto()`) and there is no guarantee that the enum values are contiguous. That's why I use a dict here to avoid potential bugs in the future. I believe there should not be any noticeable performance difference?

Reply:
It doesn't matter as you construct a concat tensor in the end anyway, right? This also ties into my other comment - if you have a list of lists, then you can create a 2d tensor with just a single torch.tensor call, which will be the most efficient

Reply:
yes the other comment makes sense. Just fixed.